File: solutions/typescript/0086.ts
/**
 * 86. 分隔链表
 * 给定链表头 head 和 x，将所有小于 x 的节点排在 >= x 的节点之前，
 * 保留各分区节点的初始相对位置。
 * 两个哑头 + 两条尾链：small 串 < x 节点，large 串 >= x 节点，最后拼接。
 *
 * 注：ListNode 类型由 LeetCode 平台注入，本文件用 untyped LNode 接口避免冲突。
 */

interface LNode { val: number; next: LNode | null }

function partition(head: LNode | null, x: number): LNode | null {
    const sDummy: LNode = { val: 0, next: null };
    const lDummy: LNode = { val: 0, next: null };
    let s: LNode = sDummy;
    let l: LNode = lDummy;
    let curr: LNode | null = head;
    while (curr) {
        const nxt = curr.next;
        curr.next = null; // 切断旧链
        if (curr.val < x) {
            s.next = curr;
            s = curr;
        } else {
            l.next = curr;
            l = curr;
        }
        curr = nxt;
    }
    s.next = lDummy.next;
    return sDummy.next;
}

export default partition;

File: solutions/rust/0086.rs
/// 86. 分隔链表
/// Box 链表 prev.next 改写复杂，与 82/83 同套路：提取值 → 分区 → 重建链表。
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     fn new(val: i32) -> Self {
//         ListNode { val, next: None }
//     }
// }

pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
    // 1. 提取所有值
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = node.next;
    }

    // 2. 稳定分区：先 < x，再 >= x
    let mut small: Vec<i32> = Vec::new();
    let mut large: Vec<i32> = Vec::new();
    for v in vals {
        if v < x {
            small.push(v);
        } else {
            large.push(v);
        }
    }
    small.extend(large);

    // 3. 重建链表
    let mut new_head: Option<Box<ListNode>> = None;
    for &v in small.iter().rev() {
        new_head = Some(Box::new(ListNode { val: v, next: new_head }));
    }
    new_head
}

impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        partition(head, x)
    }
}